feat: possibility to force coupled camera

Farenheith opened the pull request:
Some mods implement features that only work well with a coupled camera, for example, Parcool offers a way to hang on cliffs and turn around. The controls end up a little confusing with a decoupled camera. Another example is lock-on that works pretty well with a coupled camera. That way, it'd be nice if ShoulderSurfing offered a way for other mods to force a coupled camera when they need it to. This, of course, without messing up with the player configuration. In this commit I'm suggesting a way to do so.

The video below is an example of the first situation. I ended up changing shouldersurfing coupled camera configuration from this other mod to force the behavior I expected, but if this suggestion is accepted I'll be able to offer a cleaner way to solve the problem.

https://github.com/user-attachments/assets/c05ce827-b72a-4d49-ac8f-98bbb6d36e6e

Exopandora reviewed (changes requested):
Thank you for contributing! I support having such a feature. Regarding the implementation:
1. Shoulder Surfing Reloaded already supports registering thrid party plugins, which are mainly intended for compatibility resolution. I suggest moving the proposed functionality to a new callback called `ICameraCouplingCallback`, which could work similar to the `IAdaptiveItemCallback` (see api module). If you want to learn more about third party plugins, you can view the developer documentation in the [wiki](https://github.com/Exopandora/ShoulderSurfing/wiki).
2. There should be a new api method called `isCameraDecoupled()` for `IShoulderSurfing`, which replaces all occurrences of `Config.CLIENT.isCameraDecoupled()`. Its value should be computed in `ShoulderSurfingImpl.tick()`.
3. Please use the same code formatting as the current found in the codebase. If you are using IntelliJ IDEA, you can use the auto-formatter functionality with the .editorconfig file from the repository.

Farenheith commented:
@Exopandora, thanks for the review. I've already implemented the changes you suggested.

About code formatting, I'm using VSCode, actually, and I've been far from Java for more than a decade, and recently started to program again in this language due to Minecraft enthusiasm. That said, I'll take a look at how to configure my IDE appropriately.

The new commit is already available for review, but you don't need to do it right away. I'll let you know when I get linting working.

Farenheith commented on common/src/main/java/com/github/exopandora/shouldersurfing/client/ShoulderSurfingImpl.java:
```diff
@@ -114,6 +118,12 @@ else if(this.shouldEntityFollowCamera(player))
 		}
 	}
+
+	private void changeCameraDecoupling(Boolean allowDecoupling, Minecraft minecraft) {
```
```suggestion
	private void changeCameraDecoupling(Boolean allowDecoupling, Minecraft minecraft)
	{
```

Farenheith commented on common/src/main/java/com/github/exopandora/shouldersurfing/client/ShoulderSurfingImpl.java:
```diff
@@ -282,6 +297,11 @@ public boolean isAiming()
 	}
+
+	@Override
+	public boolean isCameraDecoupled() {
```
```suggestion
	public boolean isCameraDecoupled()
	{
```

Farenheith commented on common/src/main/java/com/github/exopandora/shouldersurfing/plugin/ShoulderSurfingRegistrar.java:
```diff
@@ -26,6 +28,12 @@ public IShoulderSurfingRegistrar registerAdaptiveItemCallback(IAdaptiveItemCallb
 	}
+
+	@Override
+	public IShoulderSurfingRegistrar registerCameraCouplingCallback(ICameraCouplingCallback cameraCouplingCallback) {
```
```suggestion
	public IShoulderSurfingRegistrar registerCameraCouplingCallback(ICameraCouplingCallback cameraCouplingCallback)
	{
```

Farenheith commented:
@Exopandora I've installed Intellij community and ran its formatter, on the api folder, but he changed files I haven't touched, so I'm not sure if it's what you wanted.

Regardless, I've also created a GitHub action that will run the IntelliJ formatter for each PR you receive.
If the pipeline finds anything wrong, he creates a branch and prints a link to open a PR. If you want, I can change this behavior just to notify the inconsistency or even automatically commit the fix to the origin branch. I can remove the action, too, if you don't like this suggestion.

Exopandora commented on common/src/main/java/com/github/exopandora/shouldersurfing/client/ShoulderSurfingImpl.java:
```diff
@@ -73,23 +74,26 @@ public void tick()
 		
 		this.isAiming = isHoldingAdaptiveItem(minecraft, minecraft.getCameraEntity());
 		this.updatePlayerRotations = false;
+		Boolean allowCameraDecoupling = !isFirstPerson && this.isShoulderSurfing;
 		LocalPlayer player = minecraft.player;
 		
 		if(this.isShoulderSurfing && Config.CLIENT.getCrosshairType().doSwitchPerspective(this.isAiming))
 		{
 			this.changePerspective(Perspective.FIRST_PERSON);
 			this.isTemporaryFirstPerson = true;
+			allowCameraDecoupling = false;
 		}
 		else if(this.isTemporaryFirstPerson && isFirstPerson && !Config.CLIENT.getCrosshairType().doSwitchPerspective(this.isAiming))
 		{
 			this.changePerspective(Perspective.SHOULDER_SURFING);
 		}
+		this.changeCameraDecoupling(allowCameraDecoupling, minecraft);
```
This can be simplified to:
```java
this.isCameraDecoupled = this.isShoulderSurfing && Config.CLIENT.isCameraDecoupled() && !isForcingCoupledCamera(minecraft);
```
because `this.changePerspective` will update the state of `this.isShoulderSurfing` when not in 3rd person.

Exopandora reviewed (commented):
Thanks for updating the PR! The changes look good to me. There is one spot where logic could be simplified, I think. Regarding the formatting changes:
1. It would be nice if you could revert the formatting changes of files you have not actually changed.
2. It would be nice if you could create a separete MR for the pipeline. Kudos for the effort you have put into configuring a full pipeline for linting! It might make sense to use [spotless](https://github.com/diffplug/spotless/tree/main/plugin-gradle) for linting in the future (I may have an eclipse formatter file with the correct configurations somewhere on my computer already).

Farenheith commented:
I've updated the PR with the changes requested, and I'm glad to send a new PR with the github action suggestion. You can review the code again but I appreciate if you don't merge it right away as I'm setting up my environment to test the change alongside Parcool's mod using the new feature

Farenheith commented:
@Exopandora I've tested and everything worked as expected! If you have any additional changes you want me to do, feel free to ask

Exopandora commented:
> Forgot to mention: Please use `this` wherever possible.

Once this is done and you give me your ok, I will happily merge this PR.

Farenheith commented:
> Forgot to mention: Please use `this` wherever possible.

Fixed it! I forgot about the "instance" access I had on some classes, sorry. I think I didn't forget anything else, but if you find something I missed, just let me know that I'll fix it!

Oh, and I've rebased my branch and we already can see the format pipeline running on!

Exopandora commented:
Everything looks good to me. Please let me know when i should merge it. Thanks again for contributing!

Farenheith commented:
@Exopandora everything is work as expected! It's ready to merge
